### PR TITLE
UIA: Prevent crash from invalid UTR endpoint comparison

### LINKED ISSUE #5309: Crash when attempting to compare selection endpoints with UIA
# Environment

```none
Windows build number: [run `[Environment]::OSVersion` for powershell, or `ver` for cmd] 21H1 (19603.1000)
Windows Terminal version (if applicable): inbox console

Any other software? NVDA 2020.2
```

# Steps to reproduce

1. Start [the latest NVDA alpha snapshot](http://nvda-project.org/snapshots).
2. In NVDA advanced settings, enable "use UI Automation to access the Windows Console when available".
3. Open cmd.
4. Ssh to an Ubuntu 18.04 system.
5. Run `sudo do-release-upgrade -d`.

# Expected behavior
NVDA reads the output, and the console remains functional.

# Actual behavior
Windows Console crashes, and the following is written to the NVDA log:

```none
ERROR - eventHandler.executeEvent (04:42:19.980) - MainThread (21752):
error executing event: caret on <NVDAObjects.Dynamic_WinConsoleUIAEditableTextWithAutoSelectDetectionUIA object at 0x070733B0> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 155, in executeEvent
  File "eventHandler.pyc", line 92, in __init__
  File "eventHandler.pyc", line 100, in next
  File "NVDAObjects\behaviors.pyc", line 191, in event_caret
  File "editableText.pyc", line 333, in detectPossibleSelectionChange
  File "editableText.pyc", line 340, in _updateSelectionAnchor
  File "NVDAObjects\UIA\__init__.pyc", line 776, in compareEndPoints
  File "comtypesMonkeyPatches.pyc", line 26, in __call__
_ctypes.COMError: (-2147418113, 'Catastrophic failure', (None, None, None, 0, None))
```

# Context
NVDA uses cursor changed events to detect and report selection changes to the user. It receives a cursor change event and compares the previous and current selection endpoints, which causes console to crash in this case. Is there anything I can do to help trace this on the console side?

Please push the fix for this issue to 21H1 as we'd really like to use UIA in console (especially after the excellent work in #4018 and #4495)!

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
This is a quick-and-easy solution to #5309. If the ITextRangeProvider API allows us to take in two UiaTextRanges, we need to verify that they are both valid.

With this PR, we make sure they both fit in the current TextBuffer. If not, we return `E_FAIL`. Though this doesn't prove that both UiaTextRanges are from the same TextBuffer, at the very least we don't crash and in cases where we can't make a valid comparison, we return an HRESULT failure.

## References
#5406 - This should be the proper solution to this problem. Each UiaTextRange needs to be aware of which TextBuffer it came from.

## PR Checklist
* [X] Closes #5309

## Validation Steps Performed
1. generate enough output to cause the terminal to scroll
2. execute `nano` to make us go into the alternate buffer
This previously crashed, now NVDA seems to detect that there was an error and keeps moving along.

### CODE REVIEW

**Comment by codeofdusk** [comment]
@carlos-zamora I'm assuming NVDA plays an error sound when switching into the alt mode? We *can* fix this (by just dropping that exception) but there really should be separate UIA objects for the standard and alt buffer.

**Comment by codeofdusk** [comment]
See #5406.

**Comment by carlos-zamora** [comment]
Thanks for filing that! If this is acceptable in the meantime, I'll mark it ready for review then 😊

**Comment by codeofdusk** [comment]
Could you please provide a test build?

**Comment by DHowett-MSFT** [comment]
@codeofdusk I'm working on that -- our build agents are somewhat backed up.

**Comment by DHowett-MSFT** [comment]
@codeofdusk https://1drv.ms/u/s!AtnEMxakf-D6uQ5Mn0CraCQEX4qU?e=9dEICD

File: src/types/UiaTextRangeBase.cpp [hunk 1]
```diff
@@ -227,8 +227,17 @@ try
     // get the values of our endpoint
     const auto mine = GetEndpoint(endpoint);
 
+    // TODO GH#XXXX: associate UTRs with TextBuffers
```
Comment (codeofdusk) [inline src/types/UiaTextRangeBase.cpp:230]:
```suggestion
    // TODO GH#5406: create different UIA parent objects for each TextBuffer
```

File: src/types/UiaTextRangeBase.cpp [hunk 2]
```diff
@@ -227,8 +227,17 @@ try
     // get the values of our endpoint
     const auto mine = GetEndpoint(endpoint);
 
+    // TODO GH#XXXX: associate UTRs with TextBuffers
+    //   This is a temporary solution to comparing two UTRs from different TextBuffers
```
Comment (codeofdusk) [inline src/types/UiaTextRangeBase.cpp:231]:
```suggestion
    //   This is a temporary solution to comparing two UTRs from different TextBuffers:
```

File: src/types/UiaTextRangeBase.cpp [hunk 3]
```diff
@@ -652,6 +661,17 @@ try
         return E_INVALIDARG;
     }
 
+    // TODO GH#XXXX: associate UTRs with TextBuffers
```
Comment (codeofdusk) [inline src/types/UiaTextRangeBase.cpp:664]:
```suggestion
    // TODO GH#5406: create a different UIA parent object for each TextBuffer
```

**Comment by codeofdusk** [comment]
As this fixes a crash, could this please be prioritized for inclusion in the inbox console? Thanks.

**Pull request merged**